File: Searching_Sorting/Merge_Sort/Merge_Sort.java
import java.util.*;
public class Merge_Sort {
    public static void main(String[] args) {
        int[] nums={19,1,1,1,1,1,11,1,1,1,8};
        divide(nums,0,nums.length-1);
        System.out.println(Arrays.toString(nums));
    }   
    
    public static void divide(int[] num, int start, int end){
        if(start>=end){
            return;
        }
        int mid=(start+end)/2;
        divide(num,start,mid);
        divide(num,mid+1,end);
        merge(num,start,end,mid);
    } 

    public static void merge(int[]num, int start, int end, int mid){
        int[] temp=new int[end-start+1];
        int i=start;
        int j=mid+1;
        int k=0;

        while(i<=mid && j<=end){
            if(num[i]<=num[j]){
                temp[k++]=num[i++];
            }
            else{
                temp[k++]=num[j++];
            }
        }

        while(i<=mid){
            temp[k++]=num[i++];
        }
        while(j<=end){
            temp[k++]=num[j++];
        }

        for(int x=0;x<temp.length;x++){
            num[start+x]=temp[x];
        }
    }
}

File: Searching_Sorting/Merge_Sort/Merge_Sort.py
def divide(nums,start,end):
    if start>=end:
        return
    mid=(start+end)//2
    divide(nums,start,mid)
    divide(nums,mid+1,end)
    merge(nums,start,end,mid)

def merge(nums,start,end,mid):
    i=start
    j=mid+1
    k=0
    temp=[0]*(end-start+1)

    while i<=mid and j<=end:
        if nums[i]<=nums[j]:
            temp[k]=nums[i]
            k+=1
            i+=1
        else:
            temp[k]=nums[j]
            k+=1
            j+=1

    while i<=mid:
        temp[k]=nums[i]
        k+=1
        i+=1
    while j<=end:
        temp[k]=nums[j]
        k+=1
        j+=1
    for i in range(len(temp)):
        nums[start+i]=temp[i]

nums=[1000,9,3,10,320,12,22,46]
divide(nums,0,len(nums)-1)
print(nums)